### PR TITLE
Disallow mounting filesystem entries via symlinks

### PR DESCRIPTION
**Pull request by ajeddeloh**
We can't determine the correct order to mount if we allow this. Also systemd mount units also disallow this, so there's some precedent.

### CODE REVIEW

File: tests/negative/filesystems/symlinks.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+		},
+	})
+	out := in
+	config := `{
+		"ignition": {"version": "$version"},
+		"storage": {
+			"filesystems": [{
+				"format": "ext4",
+				"path": "/tmp0",
+				"device": "/thistestdoesntneedthis"
```
Comment (arithx) [inline tests/negative/filesystems/symlinks.go]: Should we use an actual device reference so that we ensure the test causes a legitimate failure?

File: tests/negative/filesystems/symlinks.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
+// Copyright 2017 CoreOS, Inc.
```
Comment (arithx) [inline tests/negative/filesystems/symlinks.go]: Year

File: tests/negative/filesystems/symlinks.go (hunk 1 above)
Comment (jlebon) [reply tests/negative/filesystems/symlinks.go]: I'm not too familiar with the golang test framework, though for negative tests like this, I find it much more convincing to look for the error message we were expecting. Otherwise you can end up masking a regression.

File: internal/exec/stages/mount/mount.go [hunk 3]
```diff
@@ -81,11 +83,16 @@ func (s stage) mountFs(fs types.Filesystem) error {
...
+		return fmt.Errorf("Error mounting %q to %q: paths cannot include symlinks", fs.Device, fs.Path)
+	}
+
 	if _, err := os.Stat(path); err != nil && os.IsNotExist(err) {
 		if err := os.MkdirAll(path, 0755); err != nil {
 			return err
 		}
 	}
+
 	// TODO disallow symlinks
```
Comment (jlebon) [inline internal/exec/stages/mount/mount.go]: We can drop this now, right?

File: tests/negative/filesystems/symlinks.go (hunk 1 above)
Comment (ajeddeloh) [reply tests/negative/filesystems/symlinks.go]: Since we're blackbox testing we'd have to parse logs, which I'm not super keen on. I agree it's not ideal.

Comment (arithx) [reply tests/negative/filesystems/symlinks.go]: @jlebon: currently the test runner is doing a naive any failure is expected for negative testing; I'll file an issue for improving it to accept specific failure expectations.

**Comment by ajeddeloh** [comment]
Fixed the test, reworked it a little, it's clearer now and checks that the final element is a dir (that was an issue before).

File: internal/exec/stages/mount/mount.go [hunk 4]
```diff
@@ -72,12 +74,31 @@ func (s stage) Run(config types.Config) error {
...
 }
 
+// returns an error if any element of path is a symlink
+func checkForSymlinks(path string) error {
+	p := "/"
+	for _, component := range util.SplitPath(path) {
+		p = filepath.Join(p, component)
+		st, err := os.Lstat(p)
+		if err != nil && os.IsNotExist(err) {
+			return nil // nonexistant is ok
```
Comment (jlebon) [inline internal/exec/stages/mount/mount.go]: nonexistent

File: internal/exec/stages/mount/mount.go [hunk 5]
```diff
@@ -72,12 +74,31 @@ func (s stage) Run(config types.Config) error {
...
+		} else if err != nil {
+			return err
+		}
+		if !st.Mode().IsDir() {
+			return fmt.Errorf("Symlink %q specificified as part of %q. Symlinks are disallowed for mount paths", p, path)
```
Comment (jlebon) [inline internal/exec/stages/mount/mount.go]: Hmm, could be a regular file too right? Maybe reword as something like:

> `"Path %q contains non-directory component %q", path, component`

File: internal/exec/stages/mount/mount.go [hunk 6]
```diff
@@ -72,12 +74,31 @@ func (s stage) Run(config types.Config) error {
 	return nil
 }
 
+// returns an error if any element of path is a symlink
+func checkForSymlinks(path string) error {
```
Comment (jlebon) [inline internal/exec/stages/mount/mount.go]: Minor, though I'd name this `validateMountPointPath()` or something to make clearer what the intent is.

Comment (ajeddeloh) [reply internal/exec/stages/mount/mount.go]: I'd like to keep validate to mean "checks that can be done on the config without needing to be on the host". It's also not doing other validation. Maybe a comment at the caller's site instead?

Comment (jlebon) [reply internal/exec/stages/mount/mount.go]: Though now with the change above, it's no longer specifically checking for symlinks, it's just checking for "non-dirs". So maybe something like `checkDirPath`?

**Comment by cgwalters** [comment]
Related: https://github.com/systemd/systemd/pull/6293

**Comment by ajeddeloh** [comment]
@jlebon Fixed your nits, see how you like the new name for the directory check. Also realized I forgot to handle an err so that's taken care of

@cgwalters I think in this case if you're setting up a fs mount for Ignition you should know the basic structure of the system, so requiring users know where things actually are vs are linked from is fair. I'd much rather have users deal with Ignition failing than get something unexpected.

**Comment by ajeddeloh** [comment]
retest this please

**Comment by ajeddeloh** [comment]
retest this please

**Comment by ajeddeloh** [comment]
Jenkins is down for migration, tests do pass though
